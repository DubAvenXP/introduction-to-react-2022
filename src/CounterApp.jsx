import { useState } from "react";

import PropTypes from "prop-types";

// If the handleAdd function not use any prop, we can define it outside the component
// const handleAdd = (event, newValue) => {
//     console.log("Hi");
// }


// Functional component
export const CounterApp = ({ value }) => {

    // When the state changes, the component will be re-rendered with the new value
    const [counter, setCounter] = useState(value);

    const handleAdd = () => {
        setCounter(counter + 1);
        // setCounter((c) => c + 2);
    };

    const handleSubtract = () => {
        setCounter(counter - 1);
    };

    const handleReset = () => {
        setCounter(value);
    };
    return (
        <>
            <h1>CounterApp</h1>
            {/* <button onClick={ (event) => handleAdd(event, "hi") }> -1 </button> */}

            <button onClick={handleSubtract}> -1 </button>
            <span> {counter} </span>
            <button onClick={handleAdd}> +1 </button>

            <button aria-label="btn-reset" onClick={handleReset}> Reset </button>
        </>
    );
};

CounterApp.propTypes = {
    value: PropTypes.number.isRequired,
};
