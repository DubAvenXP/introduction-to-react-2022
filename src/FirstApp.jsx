import PropTypes from "prop-types";

// import {Fragment} from 'react';
// <></> is a shorthand for <Fragment></Fragment>

const newMessage = "Hola Mundo";
const arrayOfNumbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
const person = {
    name: "Juan",
    age: 25,
};

// Note: react can't render a promise or async function
const generateRandomNumber = () => Math.round(Math.random() * 100);

// /*
//     React cant render objects
//     <h3>{person}</h3>

//     React can render objects with JSON.stringify
//     <code>{JSON.stringify(person)}</code>
//     <pre>{JSON.stringify(person)}</pre>
// */

/**
 *
 * @description this ia a function component
 */
export const FirstApp = ({
    title /*  = "Hello World" */,
    subtitle /*  = 12 */,
    name,
}) => {
    // if (!title) throw new Error("Title is required");

    return (
        // <Fragment>
        // Parent node is necessary
        <>
            <p>{name}</p>
            <p>Soy un parrafo</p>
            {/* <h3>My random number {generateRandomNumber()}</h3> */}
            <h1 data-testid="test-title">{title}</h1>
            <h4 data-testid="test-subtitle">{subtitle}</h4>
            <p>{subtitle}</p>
        </>
        // </Fragment>
    );
};

FirstApp.propTypes = {
    title: PropTypes.string.isRequired,
    subtitle: PropTypes.number.isRequired,
};

FirstApp.defaultProps = {
    // title: "No title",
    // subtitle: 12,
    name: "Alejandro Dubon",
};
