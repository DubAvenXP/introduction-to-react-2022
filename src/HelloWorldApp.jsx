// /* export */ function App() {
//     return (<h1>Hello World</h1>);
// }

// export default App;

export const HelloWorldApp = () => {
    return <h1>HelloWorldApp</h1>;
};


// Other form to export a component
// export const HelloWorldApp = () => <h1>HelloWorldApp</h1>;