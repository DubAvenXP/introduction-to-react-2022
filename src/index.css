html, body {
    background-color: #21232A;
    color: #fff;
    font-family: 'Roboto', sans-serif;
    font-size: 1.3rem;
    padding: 70px;
}

span {
    color: #61dafb;
    margin: 0 30px;

}

button {
    padding: 15px;
    font-size: 1.3rem;
    border: none;
    border-radius: 5px;
    background-color: #2F80ED;
    color: #fff;
    cursor: pointer;
}