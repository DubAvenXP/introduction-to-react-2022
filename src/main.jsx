import React from "react";
import ReactDOM from "react-dom/client";

// import { HelloWorldApp } from "./HelloWorldApp";
// import { FirstApp } from "./FirstApp";
import { CounterApp } from "./CounterApp";

import "./index.css";

ReactDOM.createRoot(document.getElementById("root")).render(
    // React.StrictMode is a component that wraps the app and checks for potential problems
    <React.StrictMode>
        {/* <HelloWorldApp /> */}
        {/* <FirstApp  subtitle={123} /> */}
        <CounterApp value={10}/>
    </React.StrictMode>
);
