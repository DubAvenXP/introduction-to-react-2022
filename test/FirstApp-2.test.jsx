import { render, screen } from "@testing-library/react";
import { FirstApp } from "../src/FirstApp";

describe("Tests in <FirstApp />", () => {
    const title = "I'm Goku";
    const subtitle = 24;


    test("should be match with the snapshot", () => {
        const { container } = render(<FirstApp title={ title } subtitle={subtitle} />);
        expect(container).toMatchSnapshot();
    });


    test("should show the title in a h1 tag", () => {
        render(<FirstApp title={ title } subtitle={subtitle} />);
        expect(screen.getByRole('heading', { level: 1 }).innerHTML).toContain(title);
    })


    test("should show the subtitle sended by props", () => {
        render(<FirstApp title={ title } subtitle={subtitle} />);
        expect(screen.getAllByText(subtitle).length).toBe(2);
    });
});
