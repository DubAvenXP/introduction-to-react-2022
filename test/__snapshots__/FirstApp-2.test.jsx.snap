// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <FirstApp /> should be match with the snapshot 1`] = `
<div>
  <p>
    Alejandro Dubon
  </p>
  <p>
    Soy un parrafo
  </p>
  <h1
    data-testid="test-title"
  >
    I'm Goku
  </h1>
  <h4
    data-testid="test-subtitle"
  >
    24
  </h4>
  <p>
    24
  </p>
</div>
`;
